### PR TITLE
Jetty 12.0.x object identity ee8 and ee9

### PR DESCRIPTION
**Pull request by janbartel**
Fix to maintain object wrapper identity during cross context dispatch, only for ee8 & 9.

### CODE REVIEW

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java [hunk 1]
```diff
@@ -2501,20 +2503,115 @@ public ManagedSession getManagedSession()
...
+            return null;
+        }
+
         public void setManagedSession(ManagedSession managedSession)
         {
             _managedSession = managedSession;
+            addManagedSession(managedSession);
+        }
+
+        private void addManagedSession(ManagedSession managedSession)
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/ContextHandler.java:2546]: add some javadoc here to say that a list is maintained of all sessions from all contexts (of the same EE version) that this request has visited.

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CrossContextDispatcher.java [hunk 2]
```diff
@@ -248,6 +248,13 @@ public void forward(ServletRequest servletRequest, ServletResponse response) thr
         org.eclipse.jetty.server.Response coreResponse = coreContextRequest.getHttpChannel().getCoreResponse();
         baseResponse.resetForForward();
 
+        //if forwarding to the same environment we can take a shortcut
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CrossContextDispatcher.java]:
```suggestion
        //if forwarding to the same environment we must mutate this request for Object wrapper identity
```

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CrossContextDispatcher.java [hunk 3]
```diff
@@ -289,6 +296,13 @@ public void include(ServletRequest servletRequest, ServletResponse servletRespon
         ContextHandler.CoreContextRequest coreContextRequest = baseRequest.getCoreRequest();
         org.eclipse.jetty.server.Response coreResponse = coreContextRequest.getHttpChannel().getCoreResponse();
 
+        //if including to the same environment we can take a shortcut
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/CrossContextDispatcher.java]: ditto

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java [hunk 4]
```diff
@@ -445,29 +497,98 @@ public void setUsingUriParameters(boolean value)
...
-
-            coreRequest.setSessionManager(_sessionManager);
-            coreRequest.setRequestedSession(requestedSession);
+        try
+        {
+            switch (baseRequest.getDispatcherType())
+            {
+                case REQUEST:
+                {
+                    _sessionManager.addSessionStreamWrapper(coreRequest);
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java:518]: What if the request landed on an EE10 context first and is then forwarded to EE9.  Won't that result in the core request having a session (managed by the normal session stream wrapper) and the EE9 Request will have another session that will need to be managed by this stream wrapper?

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java [hunk 5]
```diff
@@ -66,6 +70,54 @@ public class SessionHandler extends ScopedHandler implements SessionConfig.Mutab
...
+    /**
+     * Wrapper to ensure the correct lifecycle for all sessions in all
+     * contexts to which the request has been dispatched. When a response
+     * is about to be sent back to the client, all of the sessions are
+     * given the opportunity to persist themselves. When the response is
+     * finished being handled, then the sessions have their reference
+     * counts decremented, potentially leading to eviction from their cache,
+     * as appropriate to the configuration of that cache.
+     */
+    private static class SessionStreamWrapper extends HttpStream.Wrapper
```
Comment (gregw) [inline jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java:85]: Can you say that this wrapper either replaces or is in addition to the one from the AbstractSessionManager

**Review by gregw — CHANGES_REQUESTED** [review]
Mostly javadoc changes... but I have one concern about EE10 to EE9 dispatch session management.

File: jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java (hunk 4 above)
Comment (janbartel) [reply jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/SessionHandler.java:518]: Yes, if a request created/loaded a session in an ee10 context, then that session lifecycle is managed by the core SessionStreamWrapper. If that request is then forwarded to an ee9 context, then the dispatch type is not REQUEST, so we skip adding a SessionStreamWrapper. So I think that's a problem as a session created/loaded whilst in the ee9 dispatch won't be completed as the request exists. Thinking about a solution ...

**Review by gregw — DISMISSED** [review]

**Comment by janbartel** [comment]
@gregw can I get a re-review? The jenkins failure is nothing to do with this PR, its a HTTP2 test failure.

**Review by gregw — APPROVED** [review]
CI looks like a flake

**Pull request merged**